### PR TITLE
[mobile]Configure fallback resolver for cares

### PR DESCRIPTION
**Pull request by RenjieTang**
Commit Message: Allow user to specify fallback resolvers when c-ares fails to find resolver hosts from system configs.
Risk Level: low
Testing: unit test
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: android only

### CODE REVIEW

File: mobile/library/cc/engine_builder.cc [hunk 1]
```diff
@@ -491,6 +491,10 @@ std::unique_ptr<envoy::config::bootstrap::v3::Bootstrap> EngineBuilder::generate
 #else
   if (use_cares_) {
     envoy::extensions::network::dns_resolver::cares::v3::CaresDnsResolverConfig resolver_config;
+    auto* address = resolver_config.add_resolvers();
+    address->mutable_socket_address()->set_address("8.8.8.8");
```
Comment (fredyw) [inline mobile/library/cc/engine_builder.cc:495]: I think we should also consider the IPv6 DNS addresses: https://developers.google.com/speed/public-dns/docs/using

If we are on an IPv6-only network, I don't know if adding 8.8.8.8 is going to cause any issues. We should definitely verify this first.

**Review by fredyw — CHANGES_REQUESTED** [review]

File: mobile/library/cc/engine_builder.cc (hunk 1 above)
Comment (RenjieTang) [reply mobile/library/cc/engine_builder.cc:498]: I tried configuring only `8.8.8.8` on a IPv6 only network and it worked fine. My guess is the ipv4 gets mapped to an ipv6 address somewhere.
I agree adding a ipv6 resolver feels safer though. done.

Comment (fredyw) [reply mobile/library/cc/engine_builder.cc:498]: I think having a knob is better than hardcoding it given that other users may want to use different DNS servers, such as Cloudflare, etc.

File: mobile/library/kotlin/io/envoyproxy/envoymobile/EngineBuilder.kt [hunk 2]
```diff
@@ -219,6 +220,18 @@ open class EngineBuilder() {
...
   }
 
+  /**
+   * Add fallback resolver to c_ares.
+   *
+   * @param host ip address string
+   * @param port port for the resolver
+   * @return This builder.
+   */
+  fun addFallbackResolver(host: String, port: Int): EngineBuilder {
```
Comment (fredyw) [inline mobile/library/kotlin/io/envoyproxy/envoymobile/EngineBuilder.kt:230]: nit: `caresFallbackResolver`.

File: mobile/library/jni/jni_impl.cc [hunk 3]
```diff
@@ -1326,22 +1332,23 @@ extern "C" JNIEXPORT jlong JNICALL Java_io_envoyproxy_envoymobile_engine_JniLibr
     jlong stream_idle_timeout_seconds, jlong per_try_idle_timeout_seconds, jstring app_version,
     jstring app_id, jboolean trust_chain_verification, jobjectArray filter_chain,
     jboolean enable_platform_certificates_validation, jstring upstream_tls_sni,
-    jobjectArray runtime_guards) {
+    jobjectArray runtime_guards, jobjectArray fallback_resolvers) {
```
Comment (fredyw) [inline mobile/library/jni/jni_impl.cc:1335]: nit: `cares_fallback_resolvers`

File: mobile/library/java/org/chromium/net/impl/NativeCronvoyEngineBuilderImpl.java [hunk 4]
```diff
@@ -98,6 +100,17 @@ public NativeCronvoyEngineBuilderImpl setUseCares(boolean enable) {
     return this;
   }
 
+  /**
+   * Add a fallback resolver to c_cares.
+   *
+   * @param host ip address string
+   * @param port port for the resolver
+   */
+  public NativeCronvoyEngineBuilderImpl addFallbackResolver(String host, int port) {
```
Comment (fredyw) [inline mobile/library/java/org/chromium/net/impl/NativeCronvoyEngineBuilderImpl.java:109]: nit: `addCaresFallbackResolver`.

File: mobile/library/java/io/envoyproxy/envoymobile/engine/EnvoyConfiguration.java [hunk 5]
```diff
@@ -144,7 +148,8 @@ public EnvoyConfiguration(
       List<EnvoyHTTPFilterFactory> httpPlatformFilterFactories,
       Map<String, EnvoyStringAccessor> stringAccessors,
       Map<String, EnvoyKeyValueStore> keyValueStores, Map<String, Boolean> runtimeGuards,
-      boolean enablePlatformCertificatesValidation, String upstreamTlsSni) {
+      boolean enablePlatformCertificatesValidation, String upstreamTlsSni,
+      List<Pair<String, Integer>> fallbackResolvers) {
```
Comment (fredyw) [inline mobile/library/java/io/envoyproxy/envoymobile/engine/EnvoyConfiguration.java:152]: nit: `caresFallbackResolvers`.

File: mobile/library/java/io/envoyproxy/envoymobile/engine/EnvoyConfiguration.java [hunk 6]
```diff
@@ -159,6 +164,11 @@ public EnvoyConfiguration(
     this.enableDrainPostDnsRefresh = enableDrainPostDnsRefresh;
     this.enableHttp3 = enableHttp3;
     this.useCares = useCares;
+    this.fallbackResolvers = new ArrayList<>();
+    for (Pair<String, Integer> hostAndPort : fallbackResolvers) {
```
Comment (fredyw) [inline mobile/library/java/io/envoyproxy/envoymobile/engine/EnvoyConfiguration.java:168]: It's a bit wasteful to convert from `List<Pair<String, Integer>` to `List<Pair<String, String>` and then `byte[][]`. Can we convert from `List<Pair<String, Integer>>` to `byte[][]` right away?

File: mobile/library/java/org/chromium/net/impl/NativeCronvoyEngineBuilderImpl.java [hunk 7]
```diff
@@ -50,6 +51,7 @@ public class NativeCronvoyEngineBuilderImpl extends CronvoyEngineBuilderImpl {
   private final List<String> mDnsFallbackNameservers = Collections.emptyList();
   private final boolean mEnableDnsFilterUnroutableFamilies = true;
   private boolean mUseCares = false;
+  private final List<Pair<String, Integer>> mFallbackResolvers = new ArrayList<>();
```
Comment (fredyw) [inline mobile/library/java/org/chromium/net/impl/NativeCronvoyEngineBuilderImpl.java:54]: nit: `mCaresFallbackResolvers`.

**Review by fredyw — CHANGES_REQUESTED** [review]

File: mobile/library/java/io/envoyproxy/envoymobile/engine/EnvoyConfiguration.java (hunk 6 above)
Comment (RenjieTang) [reply mobile/library/java/io/envoyproxy/envoymobile/engine/EnvoyConfiguration.java]: Yeah we use this toString pattern in multiple places and it's probably not the best solution.

We can probably follow up to clean this up. What alternatives are you thinking? Using a Bytebuffer to convert the integer?

Comment (fredyw) [reply mobile/library/java/io/envoyproxy/envoymobile/engine/EnvoyConfiguration.java]: Yeah, we can do this in a follow-up.

**Review by fredyw — APPROVED** [review]

**Review by fredyw — APPROVED** [review]

**Pull request merged**